### PR TITLE
[Fix](remote) Fix thread safety issue in cache

### LINKED ISSUE #11992: [Bug]  Thread safety issue in cache
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

master

### What's Wrong?

There may be multiple FileCache objects in the memory to open the same segment file. When one of the threads is reading the local cache file, other threads may be writing the cache file, causing the read thread to read an incomplete cache file.

### What You Expected?

work well

### How to Reproduce?

_No response_

### Anything Else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zxealous**
# Proposed changes

Issue Number: close #11992

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/io/cache/whole_file_cache.cpp [hunk 1]
```diff
@@ -60,55 +60,94 @@ Status WholeFileCache::_generate_cache_reader(size_t offset, size_t req_size) {
...
+                        // is downloaded while waiting in the thread pool
+                        RETURN_NOT_OK_STATUS_WITH_WARN(io::global_local_filesystem()->exists(
+                                                               cache_done_file, &done_file_exist),
+                                                       "Check local cache done file exist failed.");
+                    }
+                    bool cache_file_exist = false;
+                    RETURN_NOT_OK_STATUS_WITH_WARN(
+                            io::global_local_filesystem()->exists(cache_file, &cache_file_exist),
+                            "Check local cache file exist failed.");
+                    if (!done_file_exist && cache_file_exist) {
```
Comment (pengxiangyu) [inline be/src/io/cache/whole_file_cache.cpp:94]: Don't return here. Download files after delete cache_file, and do the next work in this function.

File: be/src/io/cache/whole_file_cache.cpp [hunk 2]
```diff
@@ -60,55 +60,94 @@ Status WholeFileCache::_generate_cache_reader(size_t offset, size_t req_size) {
...
+                        RETURN_NOT_OK_STATUS_WITH_WARN(
+                                io::global_local_filesystem()->delete_file(cache_file),
+                                "Check local cache file exist failed.");
+                    } else if (done_file_exist && cache_file_exist) {
```
Comment (pengxiangyu) [inline be/src/io/cache/whole_file_cache.cpp:98]: This case is better before if (!done_file_exist && cache_file_exist) {

File: be/src/io/cache/sub_file_cache.cpp [hunk 3]
```diff
@@ -121,48 +121,84 @@ Status SubFileCache::_generate_cache_reader(size_t offset, size_t req_size) {
...
+                    // is downloaded while waiting in the thread pool
+                    RETURN_NOT_OK_STATUS_WITH_WARN(io::global_local_filesystem()->exists(
+                                                           cache_done_file, &done_file_exist),
+                                                   "Check local cache done file exist failed.");
+                    bool cache_file_exist = false;
+                    RETURN_NOT_OK_STATUS_WITH_WARN(
+                            io::global_local_filesystem()->exists(cache_file, &cache_file_exist),
+                            fmt::format("Check local cache file exist failed. {}",
+                                        cache_file.native()));
+                    if (!done_file_exist && cache_file_exist) {
```
Comment (pengxiangyu) [inline be/src/io/cache/sub_file_cache.cpp:145]: Don't return here. Download files after delete cache_file, and do the next work in this function.

**Review by morningman — COMMENTED** [review]
please create an issue to describe the problem.
And add modify the document for new configs.

**Comment by zxealous** [comment]
> please create an issue to describe the problem. And add modify the document for new configs.

@morningman  done

File: be/src/io/cache/whole_file_cache.cpp (hunk 1 above)
Comment (zxealous) [reply be/src/io/cache/whole_file_cache.cpp]: done

File: be/src/io/cache/whole_file_cache.cpp (hunk 2 above)
Comment (zxealous) [reply be/src/io/cache/whole_file_cache.cpp]: done

File: be/src/io/cache/sub_file_cache.cpp (hunk 3 above)
Comment (zxealous) [reply be/src/io/cache/sub_file_cache.cpp]: done

**Review by pengxiangyu — APPROVED** [review]
LGTM

**Review by yangzhg — APPROVED** [review]
LGTM

**Pull request merged**